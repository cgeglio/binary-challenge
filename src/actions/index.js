export const login = user => ({
  type: 'LOGIN',
  user
})

export const logout = user => ({
  type: 'LOGOUT',
  user
})

export const addCards = cards => ({
  type: 'ADD_CARDS',
  cards
})

export const removeCards = cards => ({
  type: 'REMOVE_CARDS',
  cards
})

export const addQuestion = question => ({
  type: 'ADD_QUESTION',
  question
})

export const removeQuestion = question => ({
  type: 'REMOVE_QUESTION',
  question
})

export const addFortune = fortune => ({
  type: 'ADD_FORTUNE',
  fortune
})

export const removeFortune = fortune => ({
  type: 'REMOVE_FORTUNE',
  fortune
})

export const addReading = currentReading => ({
  type: 'ADD_READING',
  currentReading
})

export const removeReading = currentReading => ({
  type: 'REMOVE_READING',
  currentReading
})

export const addFavorite = favorite => ({
  type: 'ADD_FAVORITE',
  favorite
})

export const removeFavorite = favorite => ({
  type: 'REMOVE_FAVORITE',
  favorite
})

export const resetFavorites = favorites => ({
  type: 'RESET_FAVORITES',
  favorites
})
