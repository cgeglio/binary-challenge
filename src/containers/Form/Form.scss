@import "../../variables.scss";
@import "../../mixins.scss";

.login-form {
  @include flex();
  justify-content: center;
  margin-top: 410px;
  position: absolute;
  z-index: 5;
}

.login-form input {
  @include input();
  height: 30px;
  text-indent: .5em;
  width: 160px;
}

::placeholder {
  color: white;
}

.login-submit-btn {
  @include buttons();
  height: 25px;
  width: 100px;
}

.error-message {
  color: $primary-color;
  font-family: $primary-font;
  font-size: 1.5em;
  margin-top: 150px;
}
