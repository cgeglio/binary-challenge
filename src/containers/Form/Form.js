import React, { Component } from 'react';
import './Form.scss';
import { connect } from 'react-redux';
import { login } from '../../actions'


export class Form extends Component {
  constructor() {
    super()
    this.state={username: '', password: '', error: null}
  }

  handleChange = event => {
    this.setState({[event.target.name]: event.target.value})
  }

  verifyInputs = event => {
    event.preventDefault();
    let values = Object.values(this.state);
    values.includes('') ? this.setState({error: 'Enter a username & password to proceed.'}) : this.completeLogin();
  }

  completeLogin = () => {
    const id = Date.now();
    this.props.addUser({username: this.state.username, id: id, favorites: []})
    this.setState({username: '', password: '', error: null})
  }

  render() {
    return (
      <form className='login-form'>
        <input
           type='text'
           placeholder='Username...'
           value={this.state.username}
           name='username'
           onChange={this.handleChange}
        />
        <input
           type='password'
           placeholder='Password...'
           value={this.state.password}
           name='password'
           onChange={this.handleChange}
        />
        <button onClick={this.verifyInputs} className="login-submit-btn">SUBMIT</button>
        <p className="login-error-message">{this.state.error}</p>
      </form>
    )
  }
}


const mapDispatchToProps = dispatch => ({
  addUser: user => (dispatch(login(user)))
})

export default connect(null, mapDispatchToProps)(Form)
