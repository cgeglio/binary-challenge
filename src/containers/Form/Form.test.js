import React from 'react';
import { Form, mapDispatchToProps } from './Form';
import { login } from '../../actions';
import { shallow } from 'enzyme';

describe('Form', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Form />);
  });

  describe('Form container/component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should start off with a default state', () => {
      expect(wrapper.state()).toEqual({username: '', password: '', error: null});
    });

    it('should set state when handleChange is called', () => {
      const mockEvent = {
        target: {name: 'username', value: 'Bithcuits01'},
        preventDefault: jest.fn()
      }
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state()).toEqual({username: 'Bithcuits01', password: '', error: null});
    });

    it('should verify inputs when verifyInputs is called and call completeLogin', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().completeLogin = jest.fn();
      wrapper.instance().setState({username: 'Bithcuits01', password: '12345', error: null})

      wrapper.instance().verifyInputs(mockEvent)
      expect(wrapper.instance().completeLogin).toHaveBeenCalled();
    });

    it('should set state with an error if there is a missing input', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().completeLogin = jest.fn();

      wrapper.instance().verifyInputs(mockEvent)
      expect(wrapper.state('error')).toEqual('Enter a username & password to proceed.');
    });

    it('should call addUser and reset state when completeLogin is called', () => {
      global.Date.now = jest.fn().mockImplementation(() => 12345)
      const mockAddUser = jest.fn().mockImplementation();
      wrapper = shallow(<Form addUser={mockAddUser} />);
      wrapper.instance().setState({username: 'Bithcuits01', password: 'kitty', error: null});
      const mockEvent = { preventDefault: jest.fn() };

      wrapper.instance().completeLogin();
      expect(mockAddUser).toHaveBeenCalledWith({username: 'Bithcuits01',id: 12345});
      expect(wrapper.state()).toEqual({username: '', password: '', error: null});
    });

    it('should call verifyInputs when the login button is clicked', () => {
      wrapper.instance().verifyInputs = jest.fn();
      wrapper.instance().forceUpdate()
      wrapper.find('.login-submit-btn').simulate('click')
      expect(wrapper.instance().verifyInputs).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the login action when addUser is called', () => {
      const mockUser = {username: 'Bithcuits01', password: 'kitty', error: null};
      const mockDispatch = jest.fn();
      const actionToDispatch = login(mockUser);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addUser(mockUser);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });
});
