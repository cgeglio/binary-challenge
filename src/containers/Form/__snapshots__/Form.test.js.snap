// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form container/component should match the snapshot 1`] = `
<form
  className="login-form"
>
  <input
    name="username"
    onChange={[Function]}
    placeholder="Username..."
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="Password..."
    type="password"
    value=""
  />
  <button
    className="login-submit-btn"
    onClick={[Function]}
  >
    SUBMIT
  </button>
  <p
    className="error-message"
  />
</form>
`;
